src/Nats/Connection.php

```diff
@@ -443,19 +444,33 @@ public function connect($timeout = null)
         }
 
         $this->timeout      = $timeout;
-        $this->streamSocket = $this->getStream($this->options->getAddress(), $timeout);
+        $this->streamSocket = $this->getStream(
+            $this->options->getAddress(), $timeout, $this->options->getStreamContext());
         $this->setStreamTimeout($timeout);
 
-        $msg = 'CONNECT '.$this->options;
-        $this->send($msg);
-        $connectResponse = $this->receive();
+        $infoResponse = $this->receive();
 
-        if ($this->isErrorResponse($connectResponse) === true) {
-            throw Exception::forFailedConnection($connectResponse);
+        if ($this->isErrorResponse($infoResponse) === true) {
+            throw Exception::forFailedConnection($infoResponse);
         } else {
-            $this->processServerInfo($connectResponse);
+            $this->processServerInfo($infoResponse);
+            if ($this->serverInfo->isTLSRequired()) {
+                set_error_handler(
+                    function ($errno, $errstr, $errfile, $errline) {
```

Comment:
@tomponline: Should we include `restore_error_handler` here (before we throw)? Otherwise once this is caught, we won't have cleaned up the error handler we set, right?

Reply:
added